documentation/dataplane/dash-bmv2-data-plane-app.md

```diff
@@ -0,0 +1,349 @@
+- VPP port, is a veth interface and connects to BMv2 via veth pair. It serves as datapath channel to receive/send all packets between date-plane app and BMv2. Generally the port supports multi RSS queues, each queue binds to one vpp worker.
+
+**Note:** For simplicity and concept verification, vpp workers may directly call DASH SAI to offload flow to BMv2. The concern is that DASH SAI blocking API can block packet processing of vpp workers.
+
+## 6. Detailed design
+
+Referring to the below figure from [HA API HLD], it greatly outlines the whole packet flow in data plane for both standalone and HA context.
+
+![packet flow in data plane](https://github.com/sonic-net/DASH/blob/main/documentation/high-avail/images/ha-bm-packet-flow.svg)
+
+From the perspective of DPAPP, its core task in slow path is to create flow in flow table in case of flow miss in Flow (Conntrack) lookup stage, and then maintain flow state. The sub sections will depict these functions around flow object.
+
+
+### 6.1. DASH metadata
+DASH metadata records the packet processing result of DASH pipeline. It can have the following info to help flow creation/removal, etc operations:
+- flow key entry - eni_mac, vnet_id, 5-tuples
+- common flow data - state, direction, eni, flow actions, metering class, policy ID
+- flow overlay rewrite data 
+- flow underlay encapsulation data 
+- flow tunnel encapsulation data 
+
+When DASH pipeline requests DPAPP for flow creation, it encapsulates DASH metadata in an ethernet frame with EtherType DASH_METADATA and appends the original customer packet. The packet sent to DPAPP is like:
+
+<div style="text-align: center;">
+Ethernet HEADER|DASH metadata|customer packet
+</div>
+
+ The number of DASH_METADATA is 0x876D, which reuses the number of EtherType SECURE_DATA (vpp/src/vnet/ethernet/types.def at master · FDio/vpp · GitHub).
+DASH metadata is encoded in protocol DASH_METADATA, whose message format is defined as the below figure:
+
+```
+    0                   1                   2                   3   
+    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |      Source   | Type  |Subtype|           Length              |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                           Flow key entry                      |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                           Flow data                           :
+   :                                                               |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ 
+
+   Source: Where is packet from, dash_packet_source_t               
+   Type: Packet type under certain source 
+   Subtype: Packet subtype under certain type 
+   Length: Total length of DASH header
+   Flow key entry: eni_mac, 5-tuples
+   Flow data: As next figure
+```
+                                      
+```
```

Comment:
will be better to call out what does the following packet is

Reply:
claimed in `Ethernet HEADER|DASH metadata|customer packet `